src/metabase/query_processor/middleware/update_used_cards.clj

```diff
@@ -0,0 +1,45 @@
+(ns metabase.query-processor.middleware.update-used-cards
+  (:require
+   [clojure.string :as str]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.util.log :as log]
+   #_{:clj-kondo/ignore [:discouraged-namespace]}
+   [toucan2.core :as t2]))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *update-used-cards-async*
+  "Whether to update used cards asynchronously.
+  It's true by default and should only be changed for testing purposes"
+  true)
+
+(defn- update-used-cards!*
+  [used-cards]
+  (when (seq used-cards)
+    (let [f (fn []
+              (try
+                (log/debugf "%d cards were used during query execution: %s" (count used-cards) (str/join ";" (map :name used-cards)))
+                (t2/update! :model/Card :id [:in (map :id used-cards)] {:last_used_at :%now})
+                (catch Throwable e
+                  (log/error e "Error updating used cards"))))]
+      (if *update-used-cards-async*
+        (.submit clojure.lang.Agent/pooledExecutor ^Runnable f)
```

Comment:
We're doing the same thing in `process-userland-query`; should we have a centralized thread pool for this kind of operation?